Common/AnatupleProducer.py

```diff
@@ -6,29 +6,42 @@
 
 
 
-def createAnatuple(inFile, outFile, period, sample, X_mass, snapshotOptions,range, isData):
+def createAnatuple(inFile, outFile, period, sample, X_mass, snapshotOptions,range, isData, evtIds):
     Baseline.Initialize(True, True)
     df = ROOT.RDataFrame("Events", inFile)
     if range is not None:
         df = df.Range(range)
-    df = df.Define("sample", f"static_cast<int>(SampleType::{sample})")
-    df = df.Define("period", f"static_cast<int>(Period::Run{period})") 
-    df = df.Define("X_mass", f"static_cast<int>({X_mass})")
-    
-    df = df.Define("GenPart_daughters", "GetDaughters(GenPart_genPartIdxMother)")
+    if len(evtIds) > 0:
+        df = df.Filter(f"static const std::set<ULong64_t> evts = {{ {evtIds} }}; return evts.count(event) > 0;")
+    colToSave = ["event","luminosityBlock","run",
+                "MET_pt", "MET_phi","PuppiMET_pt", "PuppiMET_phi",
+                "DeepMETResolutionTune_pt", "DeepMETResolutionTune_phi","DeepMETResponseTune_pt", "DeepMETResponseTune_phi",
+                "MET_covXX", "MET_covXY", "MET_covYY", "PV_npvs","LHE_HT"] 
+ 
+    def DefineAndAppend(df, varToDefine, varToCall):
+        df = df.Define(f"{varToDefine}", f"{varToCall}")
+        colToSave.append(varToDefine)
+        return df 
+    df = DefineAndAppend(df,"sample", f"static_cast<int>(SampleType::{sample})")
+    df = DefineAndAppend(df,"period", f"static_cast<int>(Period::Run{period})") 
+    df = DefineAndAppend(df,"X_mass", f"static_cast<int>({X_mass})")
+    df = DefineAndAppend(df,"is_data", f"{isData}")   
+
     df = Baseline.RecoLeptonsSelection(df)
+    if isData==0:
+        df = df.Define("GenPart_daughters", "GetDaughters(GenPart_genPartIdxMother)")
+        df = df.Define("genLeptons","""reco_tau::gen_truth::GenLepton::fromNanoAOD(GenPart_pt, GenPart_eta,
+                                            GenPart_phi, GenPart_mass, GenPart_genPartIdxMother, GenPart_pdgId,
+                                            GenPart_statusFlags, event)""")
+        for lep in ["Electron", "Muon", "Tau"]:
+            df = DefineAndAppend(df,f"{lep}_genMatchIdx",f"is_data? MatchGenLepton({lep}_p4, genLeptons, 0.2) : defineEmptyVector({lep}_p4.size(), static_cast<int>(-1))")
```

Comment:
- move to `DefineGenObjects` in BaselineSelection.py
- change `isData` from int to bool
- Change signature of `DefineGenObjects` to:
```python
def DefineGenObjects(df, isData=False, isHH=False, Hbb_AK4mass_mpv=125.)
```
- Call it here:
```python
df = Baseline.DefineGenObjects(df, isData=isData)
```
- `*_genMatchIdx` and gen leptons should be defined independently whatever it is data or MC:
```python
if isData:
  df = df.Define("genLeptons", "std::vector<reco_tau::gen_truth::GenLepton>()")
else:
  df = df.Define("genLeptons", "reco_tau::gen_truth::GenLepton::fromNanoAOD(...)"
for lep in ["Electron", "Muon", "Tau"]:
  df = df.Define(f"{lep}_genMatchIdx",  f"MatchGenLepton({lep}_p4, genLeptons, 0.2)")
```

Reply:
+1